src/Streamly/Internal/Data/Fold.hs

```diff
@@ -384,11 +384,42 @@ transform (Pipe pstep1 pstep2 pinitial) (Fold fstep finitial fextract) =
 --
 -- | Scan the input of a 'Fold' to change it in a stateful manner using another
 -- 'Fold'.
---
--- /Unimplemented/
+-- XXX if any Fold is done assuming finish the resulting Fold?
+-- /Pre-release/
 {-# INLINE scan #-}
-scan :: Fold m a b -> Fold m b c -> Fold m a c
-scan = undefined
+scan :: Monad m => Fold m a b -> Fold m b c -> Fold m a c
+scan (Fold stepL initialL extractL) (Fold stepR initialR extractR) =
+    Fold step initial extract
+
+    where
+
+    initial = do
+        r1 <- initialL
+        r2 <- initialR
+        case r1 of
+            Done _ -> case r2 of
```

Comment:
Why are we discarding the result of first fold? We have to give it to the second fold.

Reply:
In fact the code for `initial` is usually the same as the code for `step` if you replace the "stepL (fromJust s1) x` with `initial`. So you can create a common function and pass the `action` to it. In case of initial it would be `initial` and in case of `step` it would be `stepL (fromJust s1) x` etc.